crates/control/src/search_suggestor.rs

```diff
@@ -0,0 +1,175 @@
+
+                if search_suggestion.x >= length_half {
+                    search_suggestion.x -= length_half;
+                } else {
+                    search_suggestion.x = length_half - search_suggestion.x
+                }
+                if search_suggestion.y >= width_half {
+                    search_suggestion.y -= width_half;
+                } else {
+                    search_suggestion.y = width_half - search_suggestion.y
+                }
+
+                search_suggestion.x +=
+                    1.0 / context.search_suggestor_configuration.cells_per_meter as f32 / 2.0;
+                search_suggestion.y +=
+                    1.0 / context.search_suggestor_configuration.cells_per_meter as f32 / 2.0;
+
+                suggested_search_position = Some(search_suggestion);
+            }
+        } else {
+            println!("Invalid maximum heatmap position");
+        }
+        context.heatmap.fill_if_subscribed(|| self.heatmap.clone());
+
+        Ok(MainOutputs {
+            suggested_search_position: suggested_search_position.into(),
+        })
+    }
+
+    fn update_heatmap(
+        &mut self,
+        ball_position: Option<&BallPosition>,
+        invalid_ball_positions: &Vec<HypotheticalBallPosition>,
+        robot_to_field: Option<Isometry2<f32>>,
+        cells_per_meter: usize,
+        heatmap_decay_factor: f32,
+    ) {
+        if let Some(ball_position) = ball_position {
+            if let Some(robot_to_field) = robot_to_field {
+                let ball_heatmap_position = self.calculate_heatmap_position(
+                    cells_per_meter,
+                    robot_to_field * ball_position.position,
+                );
+                if self.heatmap.get(ball_heatmap_position).is_some() {
+                    self.heatmap[ball_heatmap_position] = 1.0;
+                } else {
+                    println!("Invalid ball heatmap position");
+                }
+            }
+        }
+        for ball_hypothesis in invalid_ball_positions {
+            if let Some(robot_to_field) = robot_to_field {
+                let heatmap_position = self.calculate_heatmap_position(
+                    cells_per_meter,
+                    robot_to_field * ball_hypothesis.position,
+                );
+                if self.heatmap.get(heatmap_position).is_some() {
+                    self.heatmap[heatmap_position] =
+                        (self.heatmap[heatmap_position] + ball_hypothesis.validity) / 2.0;
+                } else {
+                    println!("Invalid hypothesis heatmap position");
+                }
+            }
+        }
+        self.heatmap = self.heatmap.clone() * (1.0 - heatmap_decay_factor);
+    }
+
+    fn calculate_heatmap_position(
+        &mut self,
+        cells_per_meter: usize,
+        hypothesis_position: Point2<f32>,
+    ) -> (usize, usize) {
+        let mut x_position: usize = 0;
```

Comment:
maybe you could create an `Isometry2` for coordinate conversion, apply that and then just convert to usize?

Reply:
Done

Reply:
After some consideration, I came to the conclusion that putting the hole heatmap into a type was the best way to deal with the coordinate transformation.